crates/server/src/certificates.rs

```diff
@@ -0,0 +1,164 @@
+use rcgen::{CertificateParams, DnType};
+use local_ip_address::local_ip;
+use std::fs;
+use std::path::Path;
+use x509_parser::prelude::*;
+use x509_parser::extensions::GeneralName;
+use std::path::PathBuf;
+
+pub fn check_for_certificate() -> eyre::Result<()> {
+    // Get the path to the certificate and key files
+    let certificate_dir = get_certificate_dir();
+    let cert_path = certificate_dir.join("cert.pem");
+    let key_path = certificate_dir.join("key.pem");
+
+    // Ensure paths exist
+    if !cert_path.exists() || !key_path.exists() {
+        generate_certificate()?;
+        write_out_instructions();
+        return Ok(());
+    }
+
+    // Check if a certificate contains local IP address
+    check_certificate()?;
+    write_out_instructions();
+    Ok(())
+}
+
+fn generate_certificate() -> eyre::Result<()> {
+    // Get the local IP address
+    let local_ip = local_ip()?;
+
+    // Generate a self-signed certificate for IP address and localhost
+    let subject_alt_names = vec![
+        local_ip.to_string(),
+        "127.0.0.1".to_string(),
+  	    "localhost".to_string()
+    ];
+
+    // Certificate name
+    let certificate_name = format!("Calimero self-signed certificate for local IP: {:?}", &local_ip.to_string());
+
+    // Create certificate parameters
+    let mut params = CertificateParams::new(subject_alt_names.clone())?;
+    // Set the distinguished name
+    params.distinguished_name.push(DnType::CommonName, certificate_name);
+
+    // Load keypair if it exists or generate a new one
+    let certificate_dir = get_certificate_dir();
+    let key_file_path = certificate_dir.join("key.pem");
+    let key_pair = if key_file_path.exists() {
+        let key_pem = fs::read_to_string(key_file_path)?;
+        rcgen::KeyPair::from_pem(&key_pem)?
+    }else {
+        rcgen::KeyPair::generate().unwrap()
+    };
+
+    // Generate the certificate with the customized parameters
+    let cert = CertificateParams::self_signed(params, &key_pair).unwrap();
+
+    // Serialize the certificate and private key to PEM format
+    let cert_pem = cert.pem();
+    let key_pem = key_pair.serialize_pem();
+
+    // Get the path to the certificate and key files
+    let certificate_dir = get_certificate_dir();
+    let cert_file_path = certificate_dir.join("cert.pem");
+    let key_file_path = certificate_dir.join("key.pem");
+    
+    // Save the certificate and key to files
+    fs::write(cert_file_path, cert_pem)?;
+    fs::write(key_file_path, key_pem)?;
```

Comment:
Consider putting these in storage instead of on-disk (non-blocker though)

You can use this for inspiration - https://github.com/calimero-network/core/blob/9896f712f887f65a7b2b62f6defd962ad28b4463/crates/server/src/admin/storage/did.rs

Reply:
For now, we agreed to leave the certificate as a file for easier installation. For future we can expand the functionality so that user can download the certificate by accessing endpoint of the server.